lib/widgets/emoji_reaction.dart

```diff
@@ -360,3 +366,261 @@ class _TextEmoji extends StatelessWidget {
+                  contentPadding: const EdgeInsetsDirectional.only(start: 10, top: 6),
+                  filled: true,
+                  fillColor: designVariables.bgSearchInput,
+                  border: OutlineInputBorder(
+                    borderRadius: BorderRadius.circular(10),
+                    borderSide: BorderSide.none),
+                  hintStyle: TextStyle(color: designVariables.textMessage)),
+                style: const TextStyle(fontSize: 19, height: 26 / 19)))),
+            TextButton(
+              onPressed: () => Navigator.pop(context),
+              style: TextButton.styleFrom(
+                padding: EdgeInsets.zero,
+                splashFactory: NoSplash.splashFactory,
+                foregroundColor: designVariables.contextMenuItemText,
+              ).copyWith(backgroundColor: WidgetStateColor.resolveWith((states) =>
+                states.contains(WidgetState.pressed)
+                  ? designVariables.contextMenuItemBg.withFadedAlpha(0.20)
+                  : Colors.transparent)),
+              child: Text(zulipLocalizations.dialogClose,
+                style: const TextStyle(fontSize: 20, height: 30 / 20))),
+          ])),
+        Expanded(child: InsetShadowBox(
+          top: 8, bottom: 8,
+          color: designVariables.bgContextMenu,
+          child: ListView.builder(
+            padding: const EdgeInsets.symmetric(vertical: 8),
+            itemCount: _resultsToDisplay.length,
+            itemBuilder: (context, i) => EmojiPickerListEntry(
+              pageContext: widget.pageContext,
+              emoji: _resultsToDisplay[i].candidate,
+              message: widget.message)))),
+    ]);
+  }
+}
+
+@visibleForTesting
+class EmojiPickerListEntry extends StatelessWidget {
+  const EmojiPickerListEntry({
+    super.key,
+    required this.pageContext,
+    required this.emoji,
+    required this.message,
+  });
+
+  final BuildContext pageContext;
+  final EmojiCandidate emoji;
+  final Message message;
+
+  static const _emojiSize = 24.0;
+  static const _notoColorEmojiTextSize = 20.1;
+
+  void _onPressed() {
+    // Dismiss the enclosing action sheet immediately,
+    // for swift UI feedback that the user's selection was received.
+    Navigator.pop(pageContext);
+
+    doAddOrRemoveReaction(
+      context: pageContext,
+      doRemoveReaction: false,
+      messageId: message.id,
+      emoji: emoji,
+      errorDialogTitle:
+        ZulipLocalizations.of(pageContext).errorReactionAddingFailedTitle);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    final designVariables = DesignVariables.of(context);
+
+    final emojiDisplay = emoji.emojiDisplay.resolve(store.userSettings);
+    final Widget? glyph = switch (emojiDisplay) {
+      ImageEmojiDisplay() =>
+        ImageEmojiWidget(size: _emojiSize, emojiDisplay: emojiDisplay),
+      UnicodeEmojiDisplay() =>
```

Comment:
```suggestion
    // TODO deduplicate this logic with [_EmojiAutocompleteItem]
    final emojiDisplay = emoji.emojiDisplay.resolve(store.userSettings);
    final Widget? glyph = switch (emojiDisplay) {
      ImageEmojiDisplay() =>
        ImageEmojiWidget(size: _emojiSize, emojiDisplay: emojiDisplay),
      UnicodeEmojiDisplay() =>
```

And then a similar comment in the other place pointing back to here.

That way at least if we start making changes in one place or the other, we'll have something to remind us that there's another copy it'd be good to keep in sync.

And also a good followup sometime would be to find a way to deduplicate the two.